File: Jump Game/Jump Game.java
public class Solution {
    public boolean canJump(int[] A) {
        if(A.length == 0)       return false;
        else if(A.length == 1)  return true;
        
        int max = A[0];
        for(int i = 0; i < A.length - 1; ++i){
            max = Math.max(max, i + A[i]);
            if(max < i + 1)         return false;
            else if(max >= A.length - 1)        return true;
        }
        
        return true;
    }
}

File: Jump Game/Jump Game.cpp
class Solution {
public:
/*
    bool recur(int A[], int n, int start){
        if(start >= n)      return true;
        
        for(unsigned int i = 1; i <= A[start]; ++i){
            if((start + i) >= n){
                return true;
            }
            else{
                if(recur(A, n, start + i)){
                    return true;
                }
            }
        }
        return false;
    }
*/    
    bool canJump(int A[], int n) {
        if(!n)      return false;
        if(n == 1)  return true;
        
//        return recur(A, n, 0);
        
        int start = A[0];
        int max = start;
        int i = 0;
        while(i < n - 1){
            start = A[i] + i;
            max = (max > start) ? max : start;
            if(max < i + 1){
                return false;
            }
            if(max >= n - 1){
                return true;
            }
            ++i;
        }
        return true;
    }
};
